index.js

```diff
@@ -119,14 +111,46 @@
   //  $$type :: String
   var $$type = '@@type';
 
-  //  type :: Any -> String
+  //  RPARSE :: RegExp
+  var RPARSE = /^(?:([^/\n\r]+)\/)?(.+?)(?:@(\d+))?$/;
+
+  //       TypeIdentifier :: (a, b, c) -> TypeIdentifier a b c
+  function TypeIdentifier(namespace, name, version) {
+    return {
+      namespace: namespace,
+      name: name,
+      version: version
+    };
+  }
+
+  //       parseTypeIdentifier :: String
+  //                           -> TypeIdentifier String? String Number?
+  function parseTypeIdentifier(s) {
+    var parsed = RPARSE.exec(s);
+    // parsed is null if the string was empty, or contained newlines
+    if (!parsed) return TypeIdentifier(null, s, null);
+    return TypeIdentifier(
+      parsed[1] || null,
+      parsed[2],
+      parsed[3] ? parseInt(parsed[3], 10) : null
+    );
+  }
+
+  //       parseNativeType :: Any -> TypeIdentifier Null String Null
+  function parseNativeType(x) {
+    var parsed = Object.prototype.toString.call(x)
+                 .slice('[object '.length, -']'.length);
+    return TypeIdentifier(null, parsed, null);
```

Comment:
The 3d line exceeds 79 characters, which is why I split it like this. I could also assign `Object.prototype.toString` to a variable.